availability/watcher.py

```diff
@@ -0,0 +1,126 @@
+SERVICE_CONN_TIMEOUT = config.get_config().get("connection_timeout", 1)
+SERVICE_READ_TIMEOUT = config.get_config().get("read_timeout", 10)
+
+LOG_LEVEL = config.get_config().get("logging", {}).get("level", "INFO")
+LOG = logging.getLogger("watcher")
+LOG.setLevel(LOG_LEVEL)
+
+
+def check_availability(data, results_queue):
+    """Check if service is available and put result to Queue.
+
+    :param data: dict that represents service query data
+    :param results_queue: Queue for results
+    """
+    try:
+        requests.get(data["url"], timeout=(SERVICE_CONN_TIMEOUT,
+                                           SERVICE_READ_TIMEOUT))
+        data["status"] = True
+    except requests.exceptions.RequestException:
+        data["status"] = False
+
+    results_queue.put(data)
+
+
+def save_availability(ts, data):
+    """Send availability data to Elastic.
+
+    :param ts: float timestamp
+    :param data: dict data to send
+    :rtype: bool
+    """
+    if not ELASTIC:
+        LOG.error("Elastic is not configured")
+        return
+    url = "%s/availability/ts/%f" % (ELASTIC, ts)
+    LOG.info("Saving: %s %r" % (url, data))
+    res = requests.put(url, data=json.dumps(data))
+    if res.status_code in (200, 201):
+        return True
+    LOG.error("Failed to save availability to Elastic (code %i): %s"
+              % (res.status_code, res.text))
+    return False
+
+
+def watch_services(period=None):
+    """Constantly check services availability.
+
+    This runs infinite services availability check with given period.
+    """
+    if not SERVICES:
+        LOG.error("No services configured. Quitting.")
+        return 1
+
+    period = period or config.get_config().get("period", 60)
+    LOG.info("Start watching with period %s seconds" % period)
+
+    if SERVICE_CONN_TIMEOUT + SERVICE_READ_TIMEOUT > period:
+        LOG.error("Period can not be lesser than timeout, "
+                  "otherwice threads could crowd round.")
+        return 1
+
+    results_queue = queue.Queue()
+
+    while True:
+        time.sleep(period)
+        for service in SERVICES:
+            LOG.info("Checking service '%(name)s' availability on %(url)s"
+                     % service)
+            data = {
+                "url": service["url"],
+                "name": service.get("name", ""),
+                "region": service.get("region", ""),
+                "time": dt.datetime.now().strftime("%Y-%m-%dT%H:%M:%S"),
+                "timestamp": time.time()
+            }
+
+            thr = threading.Thread(target=check_availability,
+                                   args=(data, results_queue))
+            thr.start()
+
+        while True:
+            try:
+                data = results_queue.get_nowait()
+                ts = data.pop("timestamp")
+                save_availability(ts, data)
+            except queue.Empty:
+                break
+
+
+if __name__ == "__main__":
+    # TODO(amaretskiy): use PID-file to prevent running
```

Comment:
we don't need to prevent this. We actually need to run many watchers at the same time with different configuration options.  For horizontal scalability, what you will do if we have 10000 URLs to check every minute?